### PR TITLE
feat: merge enhanced

### PR DESCRIPTION
**Pull request by Toricane**
## About

This pull request merges most of the features from the `enhanced` ext.
I am aiming for no breaking changes.
This will be a draft until I am finished merging the features and the documentation.

Things that are added:

1. A `Command` object, simplifying command decorators
2. Subcommand system in the `Command` object
3. `default_scope` key word argument in `Client`, letting you set a default scope for all commands (except those you disable)
4. `@option()` decorator that adds options to commands in a single line
5. Utilities such as:
    - `@autodefer()` decorator that automatically defers the command if `delay` is exceeded
    - `spread_to_rows()` helper function that spreads components into a list of `ActionRow`s automatically
6. A new way of creating an `ActionRow` (pun intended)
7. `Command.error` decorator allowing a callback coroutine when a command errors

## Checklist

- [x] I've ran `pre-commit` to format and lint the change(s) made.
- [x] I've checked to make sure the change(s) work on `3.8.6` and higher.
- [ ] This fixes/solves an [Issue](https://github.com/goverfl0w/discord-interactions/issues).
  - (If existent):
- [ ] I've made this pull request for/as: (check all that apply)
  - [x] Documentation
  - [ ] Breaking change
  - [x] New feature/enhancement
  - [ ] Bugfix

### CODE REVIEW

File: interactions/client/models/command.py [hunk 1]
```diff
@@ -181,3 +185,245 @@ class ApplicationCommand(DictSerializerMixin):
...
+            name=self.base,
+            description=self.description if self.type == 1 else None,
+            options=self.options,
+            scope=self.scope,
+            name_localizations=self.name_localizations,
+            description_localizations=self.description_localizations,
+            default_member_permissions=self.default_member_permissions,
+            dm_permission=self.dm_permission,
+        )
+        return data if isinstance(data, list) else [data]
```
Comment (sillyElvy) [inline interactions/client/models/command.py]: If you want to use this for sync it has not to be a list for global commands iirc

Comment (Toricane) [reply interactions/client/models/command.py]: Yes, I know that, I just did it for consistency and the potential to just do `.extend(data)` later on if possible.
I will tackle sync later on.

**Comment by Toricane** [comment]
Wait what I didnt mark it ready for review, how did that happen
~~how do I make it a draft again~~
done

**Comment by sillyElvy** [comment]
> Wait what I didnt mark it ready for review, how did that happen
> ~~how do I make it a draft again~~
> done

Could've been me, idk maybe the app did some shit because of my bad network

File: interactions/client/models/command.py (hunk 1 above)
Comment (Toricane) [reply interactions/client/models/command.py]: I just made it `return command(...)`, thanks

**Comment by Toricane** [comment]
Please review the code whilst I start work on documentation (dunno when that'll finish so I'd rather have the code reviews done)

File: interactions/client/bot.py [hunk 2]
```diff
@@ -493,6 +505,50 @@ async def __sync(self) -> None:  # sourcery no-metrics
...
+
+            if cmd.scope not in (MISSING, None):
+                if isinstance(cmd.scope, List):
+                    [self._scopes.add(_ if isinstance(_, int) else _.id) for _ in cmd.scope]
+                else:
+                    self._scopes.add(cmd.scope if isinstance(cmd.scope, int) else cmd.scope.id)
+
+            self.event(coro, name=f"command_{cmd.name}")
+            cmd.resolved = True
+
```
Comment (sillyElvy) [inline interactions/client/bot.py]: Imho this does not really belong into sync itself but into the _ready (if sync is off nothing would happen anyways and the command wouldn't be registered even if it is already synced)

Comment (Toricane) [reply interactions/client/bot.py]: The thing is, Extensions call __sync after being loaded by default, so having it in __sync makes sure that commands that were added from Extensions are synced properly
I *could* make it so when Extensions are loaded it appends the commands and also modifies command coroutines, but that is just more work and less dynamic

Comment (sillyElvy) [reply interactions/client/bot.py]: But if you have registered commands and disable sync the coros won't work because event is not called

Comment (Toricane) [reply interactions/client/bot.py]: Ohh I haven't tested with disabled sync
Should I make a method named something like __resolve_commands and call it both in _ready and __sync?

Comment (sillyElvy) [reply interactions/client/bot.py]: Only ready and extension load, sync should have nothing to do with it

But yeah make je method

Comment (Toricane) [reply interactions/client/bot.py]: Ok I see
I will resolve this once I've done that, thanks a lot for the review

**Comment by sillyElvy** [comment]
Also please add everything you added to the PR description, just so everyone knows

File: interactions/client/models/utils.py [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+def spread_to_rows(
+    *components: Union[ActionRow, Button, SelectMenu], max_in_row: int = 5
+) -> List[ActionRow]:  # TODO: test this out
+    """
+    docstring
+    """  # TODO: change docstring
+    if not components or len(components) > 25:
+        raise LibraryException(code=12, message="Number of components should be between 1 and 25.")
+    if not 1 <= max_in_row <= 5:
+        raise ValueError(code=12, message="max_in_row should be between 1 and 5.")
```
Comment (sillyElvy) [inline interactions/client/models/utils.py]: Ah yes, a value error with code :)

File: interactions/client/models/command.py [hunk 4]
```diff
@@ -181,3 +196,581 @@ class ApplicationCommand(DictSerializerMixin):
...
+    :type value: Optional[str]
+    :param name_localizations?: The dictionary of localization for the ``name`` field. This enforces the same restrictions as the ``name`` field.
+    :type name_localizations: Optional[Dict[Union[str, Locale], str]]
+    :param description_localizations?: The dictionary of localization for the ``description`` field. This enforces the same restrictions as the ``description`` field.
+    :type description_localizations: Optional[Dict[Union[str, Locale], str]]
+    """
+
+    def decorator(coro: Callable[..., Awaitable]) -> Callable[..., Awaitable]:
+        if isinstance(_type, int):
+            type_ = _type
```
Comment (sillyElvy) [inline interactions/client/models/command.py]: Why?

File: interactions/client/models/command.py [hunk 5]
```diff
@@ -181,3 +196,581 @@ class ApplicationCommand(DictSerializerMixin):
...
+
+    return decorator
+
+
+class StopCommand:
+    """
+    A class that when returned from a command, the command chain is stopped.
+
+    Usage: ::
+
```
Comment (sillyElvy) [inline interactions/client/models/command.py]:
```suggestion
    Usage:
    .. code-block:: python

```

File: interactions/client/models/command.py [hunk 6]
```diff
@@ -181,3 +196,581 @@ class ApplicationCommand(DictSerializerMixin):
...
+    focused: Optional[bool] = None,
+    value: Optional[str] = None,
+    name_localizations: Optional[Dict[Union[str, Locale], str]] = None,
+    description_localizations: Optional[Dict[Union[str, Locale], str]] = None,
+) -> Callable[..., Callable[..., Awaitable]]:
+    """
+    A decorator for adding options to a command.
+
+    The structure of an option: ::
+
```
Comment (sillyElvy) [inline interactions/client/models/command.py]:
```suggestion
    The structure of an option:
    .. code-block:: python 

```

File: interactions/client/models/command.py [hunk 7]
```diff
@@ -181,3 +196,581 @@ class ApplicationCommand(DictSerializerMixin):
...
+        @bot.command()
+        async def foo(ctx):
+            ... # do something
+            return StopCommand  # does not execute `bar`
+            # or `return StopCommand()`
+
+        @foo.subcommand()
+        async def bar(ctx):
+            ...  # `bar` is not executed
```
Comment (sillyElvy) [inline interactions/client/models/command.py:290]:
```suggestion
            ...  # `bar` is not executed


This allows for custom checks, for example a specific role, and just stopping the chain if the role is not present 
```

File: interactions/client/models/command.py [hunk 8]
```diff
@@ -181,3 +196,581 @@ class ApplicationCommand(DictSerializerMixin):
...
+    """
+
+
+@define()
+class BaseResult(DictSerializerMixin):
+    """
+    A class object representing the result of the base command.
+
+    Usage: ::
```
Comment (sillyElvy) [inline interactions/client/models/command.py]:
```suggestion
    Usage:
    .. code-block:: python 

```

File: interactions/client/models/command.py [hunk 9]
```diff
@@ -181,3 +196,581 @@ class ApplicationCommand(DictSerializerMixin):
...
+        return self.result
+
+
+@define()
+class GroupResult(DictSerializerMixin):
+    """
+    A class object representing the result of the base command.
+
+    Usage: ::
+
```
Comment (sillyElvy) [inline interactions/client/models/command.py]:
```suggestion
    Usage:
    .. code-block:: python 

```

**Review by sillyElvy — CHANGES_REQUESTED** [review]
More doc stuff, haven't looked that much at code yet

File: interactions/client/models/utils.py (hunk 3 above)
Comment (Toricane) [reply interactions/client/models/utils.py]: Oops need to change that thanks

File: interactions/client/models/command.py (hunk 4 above)
Comment (Toricane) [reply interactions/client/models/command.py]: 1. Don't want to use `type` as variable (do u want me to?)
2. I didn't want to reuse the variable ig

File: interactions/client/bot.py (hunk 2 above)
Comment (Toricane) [reply interactions/client/bot.py]: Done

File: interactions/client/models/command.py (hunk 4 above)
Comment (Toricane) [reply interactions/client/models/command.py]: @EdVraz pls explain

**Comment by Toricane** [comment]
> Also please add everything you added to the PR description, just so everyone knows

Done

File: interactions/client/models/command.py (hunk 4 above)
Comment (Toricane) [reply interactions/client/models/command.py]: changed it to `option_type` and simplified the entire thing

**Comment by Toricane** [comment]
Ready for review!

File: docs/quickstart.rst [hunk 10]
```diff
@@ -125,7 +125,31 @@ Now, let's look what the new parts of the code are doing:
...
+    )
+
+    @bot.command()
+    async def my_first_command(ctx: interactions.CommandContext):
+        """This is the first command I made!"""
+        await ctx.send("Hi there!")
+
+    bot.start()
+
+As of v4.3.0, you can also utilize the new command system to create commands effortlessly.
```
Comment (i0bs) [inline docs/quickstart.rst]: Move this to `migrations.rst`

File: docs/quickstart.rst [hunk 11]
```diff
@@ -166,13 +192,36 @@ Next, let's create an Option
 
     bot.start()
 
+As of v4.3.0, you can also utilize the new command system and the :ref:`@option() <models.command:Application Command Models>` decorator to create options:
```
Comment (i0bs) [inline docs/quickstart.rst]: Move this to `migrations.rst`

File: docs/quickstart.rst [hunk 12]
```diff
@@ -214,9 +263,124 @@ Nested commands: subcommands
         elif sub_command == "second_command":
           await ctx.send(f"You selected the second_command sub command and put in {second_option}")
 
+As of v4.3.0, you can also utilize the new command system to create subcommands:
```
Comment (i0bs) [inline docs/quickstart.rst]: Move this to `migrations.rst`

**Review by i0bs — CHANGES_REQUESTED** [review]
Move the new design changes to `migrations.rst` as well as clarification on that page for other features.

**Comment by Toricane** [comment]
Oh I forgot about that, thanks
Should I keep it in the quickstart and add the migration in `migration.rst`, or remove it from `quickstart.rst` and move it to `migration.rst`?

**Comment by i0bs** [comment]
> Oh I forgot about that, thanks Should I keep it in the quickstart and add the migration in `migration.rst`, or remove it from `quickstart.rst` and move it to `migration.rst`?

Migration details should only be present in migration. However, if you're trying to show alternative methods, that is fine without the need to mention a specific version.

**Review by sillyElvy — APPROVED** [review]
LGTM I think

**Review by i0bs — APPROVED** [review]

**Pull request merged**